LanguageParser.hs

```diff
@@ -1,40 +1,414 @@
+                                     P.string' "than" <* P.space1)
+
+expressionP :: Parser Expression
+expressionP = P.try varP <|>
+              P.try constantP <|>
+              P.try sumP <|>
+              P.try productP <|>
+              P.try differenceP <|>
+              P.try quotientP <|>
+              P.try squareP <|>
+              P.try cubeP <|>
+              P.try squareRootP <|>
+              P.try twiceP <|>
+              P.try modP
+
+constantP :: Parser Expression
+constantP = P.try (constP (Constant 0) (P.string' "nothing" <* P.space)) <|>
+            P.try (genericConstant (2 ^) (W.positiveNouns ++ W.neutralNouns)) <|>
+            genericConstant (negate . (2 ^)) W.negativeNouns
+            where
+              genericConstant f l =
+                Constant . f . length <$>
+                (P.try (consBeginning *> rest l) <|> rest l)
+
+              rest l = P.space *> many (oneOfString' W.adjectives <* P.space1)
+                       <* P.space <* possesives <* oneOfString' l <* P.space
+
+              possesives = optional (oneOfString' W.possesives) <* P.space
+
+              consBeginning =
+                oneOfString' (W.articles ++ W.firstPersonPossessive ++
+                             W.secondPersonPossessive ++ W.thirdPersonPossessive)
+
+              empty = const "" <$> P.takeP Nothing 0
+
+binOp :: (Expression -> Expression -> Expression) -> String -> String ->
+         Parser Expression
+binOp con word prep =
+  liftA2 con
+  (P.string' "the" *> P.space1 *> P.string' word *> P.space1 *> P.string' prep
+  *> P.space *> expressionP) (P.string' "and" *> P.space1 *> expressionP)
+
+sumP :: Parser Expression
+sumP = binOp Sum "sum" "of"
+
+differenceP :: Parser Expression
+differenceP = binOp Difference "difference" "between"
+
+productP :: Parser Expression
+productP = binOp Product "product" "of"
+
+quotientP :: Parser Expression
+quotientP = binOp Quotient "quotient" "between"
+
+unOp :: (Expression -> Expression) -> Parser a -> Parser Expression
+unOp con word = con <$> (word *> P.space *> expressionP)
+
+squareP :: Parser Expression
+squareP = unOp Square (P.string' "the" <* P.space1 <* P.string' "square" <*
+                      P.space1 <* P.string' "of")
+
+cubeP :: Parser Expression
+cubeP = unOp Cube (P.string' "the" <* P.space1 <* P.string' "cube" <*
+                  P.space1 <* P.string' "of")
+
+squareRootP :: Parser Expression
+squareRootP = unOp SquareRoot (P.string' "the" <* P.space1 <* P.string' "square"
+                              <* P.space1 <* P.string' "root" <* P.space1 <*
+                              P.string' "of")
+
+twiceP :: Parser Expression
+twiceP = unOp Twice (P.string' "twice")
+
+modP :: Parser Expression
+modP = liftA2 Mod
+       (P.string' "the" *> P.space1 *> P.string' "remainder" *> P.space1 *>
+       P.string' "of" *> P.space1 *> P.string' "the" *> P.space1 *>
+       P.string' "quotient" *> P.space1 *> P.string' "between" *> P.space1 *>
```

Comment:
absolutely want that function for this hahaha.

could be `sentenceParser "the remainder of the quotient between" *> ` instead of all this other junk.